lock-api/src/main/java/com/palantir/lock/client/BlockEnforcingLockService.java

```diff
@@ -0,0 +1,168 @@
+                BlockEnforcingLockService::clampLockRequestToDeadline,
+                this::performSingleWaitForLocksRequest,
+                response -> !response.wasSuccessful());
+    }
+
+    private static ConjureLockRequest clampLockRequestToDeadline(ConjureLockRequest request, Duration remainingTime) {
+        return ConjureLockRequest.builder()
+                .from(request)
+                .acquireTimeoutMs(Ints.saturatedCast(remainingTime.toMillis()))
+                .build();
+    }
+
+    private LockResponse performSingleLockRequest(ConjureLockRequest request) {
+        return namespacedConjureTimelockService
+                .lock(request)
+                .accept(ToLeasedLockResponse.INSTANCE);
+    }
+
+    private WaitForLocksResponse performSingleWaitForLocksRequest(ConjureLockRequest request) {
+        return ConjureLockRequests.fromConjure(namespacedConjureTimelockService.waitForLocks(request));
+    }
+
+    private enum ToLeasedLockResponse implements ConjureLockResponse.Visitor<LockResponse> {
+        INSTANCE;
+
+        @Override
+        public LockResponse visitSuccessful(SuccessfulLockResponse value) {
+            return LockResponse.successful(LeasedLockToken.of(value.getLockToken(), value.getLease()));
+        }
+
+        @Override
+        public LockResponse visitUnsuccessful(UnsuccessfulLockResponse value) {
+            return LockResponse.timedOut();
+        }
+
+        @Override
+        public LockResponse visitUnknown(String unknownType) {
+            throw new SafeIllegalStateException("Unknown response type", SafeArg.of("type", unknownType));
+        }
+    }
+
+    static class RemoteTimeoutRetryer {
+        private final Clock clock;
+
+        @VisibleForTesting
+        RemoteTimeoutRetryer(Clock clock) {
+            this.clock = clock;
+        }
+
+        static BlockEnforcingLockService.RemoteTimeoutRetryer createDefault() {
+            return new BlockEnforcingLockService.RemoteTimeoutRetryer(Clock.systemUTC());
+        }
+
+        <S, T> T attemptUntilTimeLimitOrException(
+                S request,
+                Duration duration,
+                BiFunction<S, Duration, S> durationLimiter,
+                Function<S, T> query,
+                Predicate<T> isTimedOutResponse) {
+            Instant now = clock.instant();
+            Instant deadline = now.plus(duration);
+            S currentRequest = request;
+            T currentResponse = null;
+
+            while (now.isBefore(deadline)) {
+                Duration remainingTime = Duration.between(now, deadline);
+                currentRequest = durationLimiter.apply(currentRequest, remainingTime);
+
+                try {
+                    currentResponse = query.apply(currentRequest);
+                    if (!isTimedOutResponse.test(currentResponse)) {
+                        return currentResponse;
+                    }
+                } catch (Exception e) {
```

Comment:
can this be a RuntimeException? Should not see any checked exceptions

Reply:
Yep, good point. Changed